src/entities/nestedJoin/getNestedJoinCalls.ts

```diff
@@ -0,0 +1,70 @@
+import { Token } from '../token';
+import { getPoolAddress } from '../../utils';
+import { NestedJoinInput, NestedJoinCall } from './types';
+import { NestedPoolState } from '../types';
+
+export const getNestedJoinCalls = (
+    {
+        amountsIn,
+        chainId,
+        accountAddress,
+        useNativeAssetAsWrappedAmountIn,
+        fromInternalBalance,
+    }: NestedJoinInput,
+    { pools }: NestedPoolState,
+): NestedJoinCall[] => {
+    /**
+     * Overall logic to build sequence of join calls:
+     * 1. Go from bottom pool to up filling out input amounts and output refs
+     * 2. Inputs will be amountsIn provided, output of the previous level or 0n
+     * 3. Output at max level is the bptOut
+     */
+
+    const poolsSortedByLevel = pools.sort((a, b) => a.level - b.level);
+
+    const calls: NestedJoinCall[] = [];
+    for (const pool of poolsSortedByLevel) {
+        const sortedTokens = pool.tokens
+            .sort((a, b) => a.index - b.index)
+            .map((t) => new Token(chainId, t.address, t.decimals));
+        calls.push({
+            chainId: chainId,
+            useNativeAssetAsWrappedAmountIn:
+                useNativeAssetAsWrappedAmountIn ?? false,
+            sortedTokens,
+            poolId: pool.id,
+            poolType: pool.type,
+            kind: 0,
+            sender: accountAddress,
+            recipient: accountAddress,
+            maxAmountsIn: sortedTokens.map((token) => {
```

Comment:
Can you break this logic into a separate function that is commented a bit to explain. Would make it easier to follow.
```
Inputs will be:
  - bottom pool:
    - amountsIn if provided by user
    - 0 if no user amount
  - middle pool/top pool:
    - output of the previous level
```

Reply:
That's not actually possible, because this is not entirely determined by the level of the pool (bottom vs top).
Consider the WETH/3-POOL case. The top pool has a WETH amountIn provided by the user.
I'll try to add some comments and update some variables, in order to make it easier to follow.
Let me know if that's still not enough and we'll think of something else.